File: Tracks/typescript/food-chain/food-chain.ts
export default class FoodChain {
  private static options = Object.freeze({
    fly: () => "",
    spider: (subtitle = false) =>
      (subtitle ? " that " : "\nIt ") +
      "wriggled and jiggled and tickled inside her.",
    bird: () => "\nHow absurd to swallow a bird!",
    cat: () => "\nImagine that, to swallow a cat!",
    dog: () => "\nWhat a hog, to swallow a dog!",
    goat: () => "\nJust opened her throat and swallowed a goat!",
    cow: () => "\nI don't know how she swallowed a cow!",
    horse: () => ""
  });

  private static beginning = (what: keyof typeof FoodChain.options) =>
    `I know an old lady who swallowed a ${what}.${FoodChain.options[what]()}`;

  private static middle = (
    current: keyof typeof FoodChain.options,
    previous: keyof typeof FoodChain.options,
    subtitle: string
  ) => `\nShe swallowed the ${current} to catch the ${previous}${subtitle}`;

  private static end = (what: keyof typeof FoodChain.options) =>
    "\n" +
    (what === "horse"
      ? "She's dead, of course!"
      : `I don't know why she swallowed the ${what}. Perhaps she'll die.`) +
    "\n";

  static verse = (verse: number) =>
    ([...Object.keys(FoodChain.options)] as (keyof typeof FoodChain.options)[])
      .filter((_, index) => index < verse)
      .reverse()
      .reduce(
        (verse, current, index, array) => {
          if (verse.finished === true) {
            return verse;
          }
          if (verse.phrase === "") {
            verse.phrase += FoodChain.beginning(current);
          }
          if (current === "horse" || index === array.length - 1) {
            verse.phrase += FoodChain.end(current);
            verse.finished = true;
            return verse;
          }
          if (index + 1 < array.length) {
            const next = array[index + 1];
            verse.phrase += FoodChain.middle(
              current,
              next,
              (next === "spider" && FoodChain.options[next](true)) || "."
            );
          }
          return verse;
        },
        { phrase: "", finished: false }
      ).phrase;

  static verses = (initial: number, last: number) =>
    [...Array(last - initial + 1)]
      .map((_, index) => index + initial)
      .reduce(
        (result, verseIndex) =>
          result + (result !== "" ? "\n" : "") + FoodChain.verse(verseIndex),
        ""
      );
}

File: Tracks/csharp/food-chain/FoodChain.cs
using System;
using System.Collections.Generic;
using System.Linq;

public static class FoodChain
{
    private enum Animal
    {
        fly = 1, spider, bird, cat,
        dog, goat, cow, horse
    }

    private static string AnimalToString(this Animal which) => which == Animal.spider ?
        "spider that wriggled and jiggled and tickled inside her" : Enum.GetName(typeof(Animal), which);

    private static string WhatDidSheSwallow(this Animal what) => $"I know an old lady who swallowed a {what}.";

    private static string HowDidSheSwallowIt(this Animal what) => what switch
    {
        Animal.spider => "It wriggled and jiggled and tickled inside her.",
        Animal.bird => "How absurd to swallow a bird!",
        Animal.cat => "Imagine that, to swallow a cat!",
        Animal.dog => "What a hog, to swallow a dog!",
        Animal.goat => "Just opened her throat and swallowed a goat!",
        Animal.cow => "I don't know how she swallowed a cow!",
        _ => "",
    };

    private static string WhyDidSheSwallowIt(this Animal what) => $"She swallowed the {what} to catch the {(what - 1).AnimalToString()}.";

    private static string WhatHappenedLast(this Animal what) => what == Animal.horse ?
        "She's dead, of course!" : "I don't know why she swallowed the fly. Perhaps she'll die.";

    private static bool IsLastAnimal(this Animal what) => what == Animal.fly || what == Animal.horse;

    public static string Recite(int verseNumber)
    {
        Animal currentAnimal = (Animal)verseNumber;
        var lineList = new List<string>
        {
            currentAnimal.WhatDidSheSwallow(),
            currentAnimal.HowDidSheSwallowIt()
        };
        while (!currentAnimal.IsLastAnimal())
        {
            lineList.Add(currentAnimal.WhyDidSheSwallowIt());
            currentAnimal--;
        }
        lineList.Add(currentAnimal.WhatHappenedLast());
        lineList.RemoveAll(string.IsNullOrEmpty);
        return string.Join("\n", lineList);
    }

    public static string Recite(int startVerse, int endVerse) =>
        string.Join("\n\n", Enumerable.Range(startVerse, endVerse - startVerse + 1).Select((_, index) => Recite(index + startVerse)));
}